[JSC] Close Iterator Helpers underlying iterator when arguments are invalid

sosukesuzuki opened the pull request:
#### 6e8fef40092d06a84de870150952f169c94bd953
<pre>
[JSC] Close Iterator Helpers underlying iterator when arguments are invalid
<a href="https://bugs.webkit.org/show_bug.cgi?id=284709">https://bugs.webkit.org/show_bug.cgi?id=284709</a>

Reviewed by Yusuke Suzuki.

The normative change[1] that requires closing underlying iterators when
argument validation fails for iterator helpers reached consensus at the
TC39 meeting in December 2024[2].

This patch implements it.

[1]: <a href="https://github.com/tc39/ecma262/pull/3467">https://github.com/tc39/ecma262/pull/3467</a>
[2]: <a href="https://github.com/tc39/agendas/blob/main/2024/12.md">https://github.com/tc39/agendas/blob/main/2024/12.md</a>

* JSTests/stress/iterator-helpers-close-for-invalid-argument.js: Added.
(shouldThrow):
(shouldBe):
(throw.new.Error.let.closable.get next):
(throw.new.Error):
(shouldBe.let.closable.get next):
(shouldBe.OurError):
(let.closable.get next):
(shouldBe.get shouldBe):
(OurError):
(get shouldBe):
* Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
(some.wrapper.iterator):
(some):
(every.wrapper.iterator):
(every):
(find.wrapper.iterator):
(find):
(reduce):

Canonical link: <a href="https://commits.webkit.org/290907@main">https://commits.webkit.org/290907@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9adf7e2d10c0ff0fa72c0186c2bc71b8a1ec9cde

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88634 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8153 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93596 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39388 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8540 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16337 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68339 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26042 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6537 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80159 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48704 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6300 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34569 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38496 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81433 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35468 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95434 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87410 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15809 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11571 "Found 7 new test failures: http/tests/webarchive/cross-origin-stylesheet-crash.html http/tests/webarchive/test-css-url-encoding-shift-jis.html http/tests/webarchive/test-css-url-encoding-utf-8.html http/tests/webarchive/test-css-url-encoding.html http/tests/webarchive/test-preload-resources.html imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-peerconnection.https.html pdf/pdf-plugin-printing.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77203 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76018 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76479 "Found 2 new API test failures: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy, /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19400 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20867 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19290 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8848 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15825 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21133 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109903 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15566 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26412 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19015 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17348 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -278,12 +307,15 @@ function reduce(reducer /*, initialValue */)
         @throwTypeError("Iterator.prototype.reduce requires that |this| be an Object.");
 
-    if (!@isCallable(reducer))
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -256,12 +282,15 @@ function find(predicate)
         @throwTypeError("Iterator.prototype.find requires that |this| be an Object.");
 
-    if (!@isCallable(predicate))
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -212,12 +233,14 @@ function some(predicate)
         @throwTypeError("Iterator.prototype.some requires that |this| be an Object.");
 
-    if (!@isCallable(predicate))
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -179,10 +197,13 @@ function flatMap(mapper)
         @throwTypeError("Iterator.prototype.flatMap requires that |this| be an Object.");
 
-    if (!@isCallable(mapper))
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -138,15 +150,21 @@ function drop(limit)
 
-    var numLimit = @toNumber(limit);
-    if (numLimit !== numLimit)
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -94,15 +100,21 @@ function take(limit)
 
-    var numLimit = @toNumber(limit);
-    if (numLimit !== numLimit)
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -63,10 +66,13 @@ function filter(predicate)
         @throwTypeError("Iterator.prototype.filter requires that |this| be an Object.");
 
-    if (!@isCallable(predicate))
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
```diff
@@ -32,10 +32,13 @@ function map(mapper)
         @throwTypeError("Iterator.prototype.map requires that |this| be an Object.");
 
-    if (!@isCallable(mapper))
+    var iterated = this;
```
Let's just use `this` directly in this function instead of `iterated` when possible.

Constellation reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Close Iterator Helpers underlying iterator when arguments are invalid
https://bugs.webkit.org/show_bug.cgi?id=284709

Reviewed by Yusuke Suzuki.

The normative change[1] that requires closing underlying iterators when
argument validation fails for iterator helpers reached consensus at the
TC39 meeting in December 2024[2].

This patch implements it.

[1]: https://github.com/tc39/ecma262/pull/3467
[2]: https://github.com/tc39/agendas/blob/main/2024/12.md

* JSTests/stress/iterator-helpers-close-for-invalid-argument.js: Added.
(shouldThrow):
(shouldBe):
(throw.new.Error.let.closable.get next):
(throw.new.Error):
(shouldBe.let.closable.get next):
(shouldBe.OurError):
(let.closable.get next):
(shouldBe.get shouldBe):
(OurError):
(get shouldBe):
* Source/JavaScriptCore/builtins/JSIteratorPrototype.js:
(some.wrapper.iterator):
(some):
(every.wrapper.iterator):
(every):
(find.wrapper.iterator):
(find):
(reduce):

Canonical link: https://commits.webkit.org/290907@main
